### PR TITLE
feat: add Singleton class support

### PR DESCRIPTION
**Pull request by willemneal**
This PR aims allows a single class to be exported from the main file.  This is the same functionality when using the [rust-sdk](https://github.com/near/near-sdk-rs). 

Like in sdk-rs, there is a special key "STATE", which stores the instance fields of the singleton class.  At the top level there is a global called "__contract", which is initial undefined (null).  However, if there is an entry for "STATE" in storage it is used to define __contract.  Each public method is transformed into an exported function. The constructor is exported as "new" and checks to ensure that it can only be initialized once.  The other methods check to ensure that the contract has been initialized.

```ts
import { logging, PersistentSet, Context, persist } from "near-sdk-as"

@nearBindgen
export class Singleton {
  private visitors: PersistentSet<string> = new PersistentSet("v");
  private _owner: string
  private last_visited: string = "NULL";
  constructor(owner: string){
    this._owner = owner;
  };

  owner(): string {
    return this._owner;
  }

  @mutateState()
  visit(): void {
    if (!this.visitors.has(Context.sender)) {
      logging.log("Visited the first time by " + Context.sender);
      this.visitors.add(Context.sender);
    }
    this.last_visited = Context.sender;
  }

  visit_without_updated_decorator(): void {
    this.visit();
    persist(this);
  }

  visit_without_change(): void {
    this.visit();
  }

  hasVisited(visitor: string): bool {
    return this.visitors.has(visitor);
  }

  lastVisited(): string {
    return this.last_visited || "";
  }

  private hasNotVisited(visitor: string): boolean {
    return !this.hasNotVisited(visitor);
  }
} 
```
becomes

```ts
// Previous class no longer exported

let __contract: Singleton
if (__checkState()) {
  __contract = __getState<Singleton>();
}
function init(owner: string): void {
  assert(isNull(__contract), "contract is already initialized");
  __contract = new Singleton(owner);
  __setState(__contract);
}

function owner(): string {
  assert(!isNull(__contract), "contract is not initialized");
  let res = __contract.owner();
  return res;
}

export function visit(): void {
  assert(!isNull(__contract), "contract is not initialized");
  __contract.visit();
  __setState(__contract);
}
export function visit_without_updated_decorator(): void {
  assert(!isNull(__contract), "contract is not initialized");
  __contract.visit_without_updated_decorator();
}
export function visit_without_change(): void {
  assert(!isNull(__contract), "contract is not initialized");
  __contract.visit_without_change();
}
function hasVisited(visitor: string): bool {
  assert(!isNull(__contract), "contract is not initialized");
  let res = __contract.hasVisited(visitor);
  return res;
}
function lastVisited(): string {
  assert(!isNull(__contract), "contract is not initialized");
  let res = __contract.lastVisited();
  return res;
}

@exportAs("new")
function __wrapper_init(): void {
  const obj = getInput();
  init(obj.has('owner') ? 
             decode<string, JSON.Obj>(obj, "owner") : requireParameter<string>("owner"));
}
export { __wrapper_init as init }
function __wrapper_owner(): void {
  let result: string = owner();
  const val = encode<string>(result);
  value_return(val.byteLength, val.dataStart);
}
export { __wrapper_owner as owner }
function __wrapper_hasVisited(): void {
  const obj = getInput();
  let result: bool = hasVisited(obj.has('visitor') ? 
             decode<string, JSON.Obj>(obj, "visitor") : requireParameter<string>("visitor"));
  const val = encode<bool>(result);
  value_return(val.byteLength, val.dataStart);
}
export { __wrapper_hasVisited as hasVisited }
function __wrapper_lastVisited(): void {
  let result: string = lastVisited();
  const val = encode<string>(result);
  value_return(val.byteLength, val.dataStart);
}
export { __wrapper_lastVisited as lastVisited }
```
~In rust the instance variable `&self` is by default immutable and `&mut self` is mutable.  Thus in sdk-rs to check if a function is a view function, the mutability of `self` is checked.  Since AS doesn't allow for this, instead there is a new mutable global called `__updated` which is initially false.  Then each field in the class is renamed and the old name is used as a read/write property.  Any use of a setter sets `__updated` to true.  `__updateState` then conditionally checks if the contract was mutated and serializes and writes the class to storage.~

### CODE REVIEW

**Review by chadoh — CHANGES_REQUESTED** [review]
This seems like a big change! Will old contracts continue to work, or will they need to be rewritten? If they need to be rewritten, what is the migration path?

The description of this PR is decent, but I don't see any similar documentation changes within the changed files at all. This seems like very core behavior that we need to explain as soon as people click on a link to learn more about near-sdk-as. I want to make sure we ship this with copious & understandable documentation, which will be available both in the root README at https://near.github.io/near-sdk-as/ as well as inline in relevant files. 

Some questions/comments about the behavior you described:

* `The constructor is exported as "new"` – I think I mostly like this design decision, though it does break with near-sdk-rs, which uses the `#[init]` decorator and doesn't care what the function is called. I like that `near-sdk-rs` makes the function name explicit, though I appreciate that I can use a constructor that feels more familiar coming from JavaScript.
* If I have a `new` method in my class, will a useful error be thrown?
* Can I leave out the `constructor`?
* `instead there is a new mutable global called __updated` – check with @evgenykuzyakov if this will give us the behavior we need to replicate functionality from Berry Farm. It is my current understanding that the `touch` method [here](https://github.com/evgenykuzyakov/berryfarm/blob/master/contract-rs/berry-farm/src/lib.rs) will mutate contract state if called from a change method, but when called from a view function will make similar changes which are not persisted.

**Comment by willemneal** [comment]
Firstly this shouldn't impact previous contracts; it only adds the ability to export a class (which while possible before, wouldn't play well and with most examples using a model class in a separate file wasn't encouraged).  

> - The constructor is exported as "new" - I think I mostly like this design decision, though it does break with near-sdk-rs, which uses the #[init] decorator and doesn't care what the function is called. I like that near-sdk-rs makes the function name explicit, though I appreciate that I can use a constructor that feels more familiar coming from JavaScript.

This should've been clearer, but you can still choose to export the constructor as any name `@exportAs("init")` etc.

> - If I have a new method in my class, will a useful error be thrown?

Good test case to add.

> - Can I leave out the constructor?

This should mean there is no actual constructor function so it should be fine; again another good test case.

> - instead there is a new mutable global called __updated – check with @evgenykuzyakov if this will give us the behavior we need to replicate functionality from Berry Farm. It is my current understanding that the touch method here will mutate contract state if called from a change method, but when called from a view function will make similar changes which are not persisted.

This last one I am going to change.  It's not well thought out (just a hack) and after some thought it seems reasonable for users to have to add a `@change` decorator, which then requires the state to be written back.  This would solve the issue you bring up for berry club.

**Comment by amgando** [comment]
@willemneal seems like a line from this change leaked into a published version

https://github.com/figment-networks/tutorials/issues/5

**Comment by willemneal** [comment]
@chadoh 

In this example: `updateState` is a global decorator, which updates the state of the contract after the function call (same as `change` before), but there is also a `commit` method that can be imported that you can manually use to write the state back to storage.  These both can be renamed of course, but just wanted your thoughts on the usage.


```ts
import { logging, PersistentSet, Context, commit } from "near-sdk-as";


@nearBindgen
export class Singleton {

  private visitors: PersistentSet<string> = new PersistentSet("v");
  private _owner: string
  private last_visited: string = "NULL";
  constructor(owner: string){
    this._owner = owner;
  };

  owner(): string {
    return this._owner;
  }

  // Currently the `()` is needed because of TypeScript  to make the IDE happy...
  @updateState()
  visit(): void {
    if (!this.visitors.has(Context.sender)) {
      logging.log("Visited the first time by " + Context.sender);
      this.visitors.add(Context.sender);
    }
    this.last_visited = Context.sender;
  }

  visit_without_updated_decorator(): void {
    this.visit();
    commit(this);
  }

  visit_without_change(): void {
    this.visit();
  }

  hasVisited(visitor: string): bool {
    return this.visitors.has(visitor);
  }

  lastVisited(): string {
    return this.last_visited || "";
  }

  private hasNotVisited(visitor: string): boolean {
    return !this.hasNotVisited(visitor);
  }
}  
```

**Comment by chadoh** [comment]
I like the explicit `import` and use of a function much more than the decorator approach! 

`commit` is _okay_ for a name. Other names I can think of:

* `persist`
* `save`
* `write`
* `record`
* `updateContractStorage`

**Comment by amgando** [comment]
> I like the explicit `import` and use of a function much more than the decorator

I can see both being useful 

- some decorator could serve at function scope to simply declare that when the function scope exits, state will be updated...whatever then specific text is, maybe `@mutate` or `@mutateState` could bring this into alignment with Rust jargon
- some explicit import and invocation would serve well in cases where an assertion can panic but you want to be sure to capture some state regardless, before the possible panic. Not sure if it would work that way or panic rolls back all changes but just a thought

File: simulator/assembly/__tests__/singleton-fail.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return this.visitors.has(visitor);
+  }
+
+  lastVisited(): string {
+    return this.last_visited || "";
+  }
+
+  private hasNotVisited(visitor: string): boolean {
+    return !this.hasNotVisited(visitor);
+  }
```
Comment (chadoh) [inline simulator/assembly/__tests__/singleton-fail.ts:40]: Are any of these methods used? Seems like this file is only here to test that the `new` function below causes errors. If that's true, let's get rid of everything except the code under test.

File: simulator/__tests__/singleton.spec.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  return new Promise(asb);
+}
+
+describe("Complier fails", () => {
+  it("shouldn't allow methods with the same name as init function", async () => {
+    try {
+      await compile("singleton-fail")
+      expect(true).toBe(false);
+    } catch (e) {
+      expect(e.message).toContain(`Method "new" already used cannot export constructor using the same name.`);
```
Comment (chadoh) [inline simulator/__tests__/singleton.spec.ts:40]: Some extra punctuation in this message would be nice 🙂

> Method "new" already used; cannot export constructor using the same name

File: simulator/__tests__/singleton.spec.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    }
+  })
+})
```
Comment (chadoh) [inline simulator/__tests__/singleton.spec.ts:43]: This seems like a good place to test that you can use the `@nearBindgen` decorator on a class with no `constructor`

File: simulator/__tests__/singleton.spec.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      "singleton",
+      __dirname + "/../build/debug/singleton.wasm"
+    );
+  });
+
+  function init() {
+    return alice.call_other("singleton", "new", { owner: "alice" });
+  }
+
+  it("should only initialze once", () => {
```
Comment (chadoh) [inline simulator/__tests__/singleton.spec.ts:60]: typo

**Review by chadoh — CHANGES_REQUESTED** [review]
Some small ideas for improvements below. I'd mark as "approve" except that I don't see "@nearBindgen used with class that has no constructor" being tested, which seems like an important edge case to cover before we release this. Probably quick to add?

File: simulator/assembly/__tests__/singleton-fail.ts (hunk 1 above)
Comment (willemneal) [reply simulator/assembly/__tests__/singleton-fail.ts:40]: This method shouldn't end up on the final binary's exports.  I didn't have a test for this just checked  the output myself.  I'll add a test.

Comment (willemneal) [reply simulator/assembly/__tests__/singleton-fail.ts:40]: nevermind I do have a test for this:

```ts
  it("should not have private methods", () => {
    init()
    let res = alice.call_other("singleton", "hasNotVisited", {});
    expect(res.err["FunctionCallError"]["MethodResolveError"]).toContain("MethodNotFound");
  })
```

**Comment by willemneal** [comment]
@chadoh 

So no constructor turned out to be a bit more complicated that I had initially thought.  If there is no constructor than there is no place to ensure that the contract has been initialized.  So the generated top level code looks like the following in the case of no constructor:
```ts
let __contract: Singleton
if (__checkState()) {
  __contract = __getState<Singleton>();
} else {
  __contract = new Singleton();
}
function setOwner(owner: string): void {
  __contract.setOwner(owner);
  __setState(__contract);
}
function owner(): string {
  let res = __contract.owner();
  return res;
}

...
```

This doesn't check if the contract is initialized as there is a default initialization in the case where it hasn't been written to storage.

File: simulator/assembly/__tests__/singleton-fail.ts (hunk 1 above)
Comment (chadoh) [reply simulator/assembly/__tests__/singleton-fail.ts:40]: Right, but that uses a different file. Confusing, huh? 😝

**Review by chadoh — APPROVED** [review]

**Pull request merged**